docs/rfc/rfc-002-zero-copy-encoding.md

```diff
@@ -0,0 +1,406 @@
+pointers to a variable-length value:
+`string y = 2 [(cosmos_proto.fixed_size) = 3]` could indicate that this is a fixed width 3 byte string
+
+If we choose to enable these encoding options, changing these options would be a breaking change that needs to be
+prevented by the breaking change detector.
+
+### Generated Code
+
+We will describe the generated Go and Rust code using this example protobuf file:
+
+```protobuf
+message Foo {
+  int32 x = 1;
+  optional uint32 y = 2;
+  string z = 3;
+  Bar bar = 4;
+  repeated Bar bars = 5;
+}
+
+
+message Bar {
+  ABC abc = 1;
+  Baz baz = 2;
+  repeated uint32 xs = 3;
+}
+
+message Baz {
+  oneof sum {
+    uint32 x = 1;
+    string y = 2;
+  }
+}
+
+enum ABC {
+  A = 0;
+  B = 1;
+  C = 2;
+  D = 3;
+}
+```
+
+#### Go
+
+In golang, the generated code would not expose any exported struct fields, but rather getters and setters as an
+interface
+or struct methods, ex:
+
+```go
+type Foo interface {
+    X() int32
+    SetX(int32) 
+    Y() zpb.Option[uint32]
+    SetY(zpb.Option[uint32])
+    Z() (string, error)
+    SetZ(string) error
+    Bar() Bar
+    Bars() (zpb.Array[Bar], error)
+}
+
+type Bar interface {
+    Abc() ABC
+    SetAbc(ABC) Bar
+    Baz() Baz
+    Xs() (zpb.ScalarArray[uint32], error)
+}
+
+type Baz interface {
+    Case() Baz_case
+    GetX() uint32
+    SetX(uint32)
+    GetY() (string, error)
+    SetY(string)
+}
+
+type Baz_case int32
+const (
+    Baz_X Baz_case = 0
+    Baz_Y Baz_case = 1
+)
+
+type ABC int32
+const (
+    ABC_A ABC = 0
+    ABC_B ABC = 1
+    ABC_C ABC = 2
+    ABC_D ABC = 3
+)
+```
+
+Special types `zpb.Option`, `zpb.Array` and `zpb.ScalarArray` are used to represent `optional` and repeated fields
+respectively. These types would be included in the runtime library (called `zpb` here for zero-copy protobuf) and would
+have an API like this:
+
+```go
+type Option[T] interface {
+    IsSet() bool
+    Value() T
+}
+
+type Array[T] interface {
+    InitWithLength(int) error
+    Len() int
+    Get(int) T
+}
+
+type ScalarArray[T] interface {
+    Array[T]
+    Set(int, T)
+}
+```
+
+Arrays in particular would not be resizable, but would be initialized with a fixed length. This is to ensure that arrays
+can be written to the underlying buffer in a linear way.
+
+In golang, buffers would be managed transparently under the hood by the first message initialized, and usage of this
+generated code might look like this:
+
+```go
+foo := NewFoo()
+foo.SetX(1)
+foo.SetY(zpb.NewOption[uint32](2))
+err := foo.SetZ("hello")
+if err != nil {
+    panic(err)
+}
+
+bar := foo.Bar()
+bar.Baz().SetX(3)
+
+xs, err = bar.Xs()
+if err != nil {
+    panic(err)
+}
+xs.InitWithLength(2)
+xs.Set(0, 0)
+xs.Set(1, 2)
+
+bars, err = foo.Bars()
+if err != nil {
+    panic(err)
+}
+bars.InitWithLength(3)
+bars.Get(0).Baz().SetY("hello")
+bars.Get(1).SetAbc(ABC_B)
+bars.Get(2).Baz().SetX(4)
```

Comment:
Methods like InitWithLength, Set, and even Get imply that the behavior of those methods is expressed purely as side effects on the underlying values. And, since they don't seem to return anything, the result of that behavior seems to be unknowable to callers.

A method's return value is not expected to be mutable by callers in the way that is implied here. This is a highly non-idiomatic API. It shouldn't be an assumption of any higher-level design.

Reply:
That's a valid point. We should probably think of a better API